modules/internal/DefaultAssociative.chpl

```diff
@@ -566,14 +566,15 @@ module DefaultAssociative {
 
     // ref version
     proc dsiAccess(idx : idxType, haveLock = false) ref {
-      const shouldLock = dom.parSafe && !haveLock;
+      const arrOwnsDom = dom._arrs.length == 1;
+      const shouldLock = dom.parSafe && !haveLock && arrOwnsDom;
```

Comment:
I'm confused:

* It isn't a shorthand for the code you wrote here, but surely it's a shorthand for something...?  
* In any case, I'm not sure what you're concluding...

Since it doesn't seem that we're disagreeing that the "lock only when a write access misses and needs to grow things" isn't the right solution, I've pushed a new commit that locks more aggressively on write dsiAccesses when the 1:1 property exists as a placeholder for the time being. 

I also think we're going to need @benharsh (if not others) to look back in on this PR and weigh in.